File: ingest-rest/src/main/java/org/chronopolis/ingest/models/RegionCreate.java
package org.chronopolis.ingest.models;

import org.chronopolis.rest.models.storage.DataType;
import org.chronopolis.rest.models.storage.StorageType;
import org.hibernate.validator.constraints.NotEmpty;

import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;

/**
 * Data required when creating a storage region
 *
 * Created by shake on 7/11/17.
 */
public class RegionCreate {

    @NotNull
    private String node;

    @NotNull
    @Min(1L)
    private Long capacity;

    @NotNull
    private DataType dataType;

    @NotNull
    private StorageType storageType;

    // Subtype using the ReplicationCreate? (explore after)
    private String replicationUser;

    @NotNull
    @NotEmpty
    private String replicationPath;

    @NotNull
    @NotEmpty
    private String replicationServer;

    public Long getCapacity() {
        return capacity;
    }

    public RegionCreate setCapacity(Long capacity) {
        this.capacity = capacity;
        return this;
    }

    public StorageType getStorageType() {
        return storageType;
    }

    public RegionCreate setStorageType(StorageType storageType) {
        this.storageType = storageType;
        return this;
    }

    public String getReplicationUser() {
        return replicationUser;
    }

    public RegionCreate setReplicationUser(String replicationUser) {
        this.replicationUser = replicationUser;
        return this;
    }

    public String getReplicationPath() {
        return replicationPath;
    }

    public RegionCreate setReplicationPath(String replicationPath) {
        this.replicationPath = replicationPath;
        return this;
    }

    public String getReplicationServer() {
        return replicationServer;
    }

    public RegionCreate setReplicationServer(String replicationServer) {
        this.replicationServer = replicationServer;
        return this;
    }

    public String getNode() {
        return node;
    }

    public RegionCreate setNode(String node) {
        this.node = node;
        return this;
    }

    public DataType getDataType() {
        return dataType;
    }

    public RegionCreate setDataType(DataType dataType) {
        this.dataType = dataType;
        return this;
    }
}

File: rest-models/src/main/kotlin/org/chronopolis/rest/models/create/RegionCreate.kt
package org.chronopolis.rest.models.create

import org.chronopolis.rest.models.Replication
import org.chronopolis.rest.models.StorageRegion
import org.chronopolis.rest.models.enums.DataType
import org.chronopolis.rest.models.enums.StorageType
import org.chronopolis.rest.models.enums.StorageUnit
import org.hibernate.validator.constraints.NotEmpty
import javax.validation.constraints.Min

/**
 * A request to create a [StorageRegion]
 *
 * @property note a note to further identify the [StorageRegion]
 * @property node the Chronopolis Node which the [StorageRegion] belongs to
 * @property capacity the storage capacity of the [StorageRegion], used with the [storageUnit]
 * @property dataType the [DataType] identifying what the [StorageRegion] is used for
 * @property storageUnit the [StorageUnit] used to conjunction with the [capacity]
 * @property storageType the [StorageType] defining what type of storage this is
 * @property replicationPath the path to be used when creating [Replication]s
 * @property replicationServer the server to be used when creating [Replication]s
 * @property replicationUser the user to use when creating [Replication]s, or null
 * @author shake
 */
data class RegionCreate(var note: String = "",
                        @get:NotEmpty var node: String = "",
                        @get:Min(1) var capacity: Long = 0,
                        var dataType: DataType = DataType.BAG,
                        var storageUnit: StorageUnit = StorageUnit.B,
                        var storageType: StorageType = StorageType.LOCAL,
                        @get:NotEmpty var replicationPath: String = "",
                        @get:NotEmpty var replicationServer: String = "",
                        var replicationUser: String? = null) {

    /**
     * Return a normalized representation of the [capacity] by multiplying it by 1000 raised to the
     * power of the [storageUnit]
     */
    fun normalizedCapacity(): Double {
        return capacity * Math.pow(1000.0, storageUnit.power.toDouble())
    }
}